### PR TITLE
Qos 3 - Add QoS policy types and conversions

### PR DESCRIPTION
**Pull request by emersonknapp**
Add new Lifespan, Deadline, and Liveliness QoS policies - with their conversions to and from C

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: rclpy/src/rclpy/_rclpy.c [hunk 1]
```diff
@@ -3660,6 +3668,18 @@ rclpy_convert_from_py_qos_policy(PyObject * Py_UNUSED(self), PyObject * args)
   qos_profile->depth = pyqos_depth;
   qos_profile->reliability = pyqos_reliability;
   qos_profile->durability = pyqos_durability;
+
+  qos_profile->lifespan.sec = RCUTILS_NS_TO_S(pyqos_lifespan);
+  qos_profile->lifespan.nsec = pyqos_lifespan - RCUTILS_NS_TO_S(qos_profile->lifespan.sec);
```
Comment (nburek) [inline rclpy/src/rclpy/_rclpy.c]:
```suggestion
  qos_profile->lifespan.nsec = pyqos_lifespan - RCUTILS_S_TO_NS(qos_profile->lifespan.sec);
```

You need to go from Sec to NSec. Looks like this is already correct on the other lines below.

File: rclpy/src/rclpy_common/src/common.c [hunk 2]
```diff
@@ -75,6 +103,11 @@ rclpy_convert_to_py_qos_policy(void * profile)
   set_result += PyObject_SetAttrString(pyqos_profile, "history", pyqos_history);
   set_result += PyObject_SetAttrString(pyqos_profile, "reliability", pyqos_reliability);
   set_result += PyObject_SetAttrString(pyqos_profile, "durability", pyqos_durability);
+  set_result += PyObject_SetAttrString(pyqos_profile, "lifespan", pyqos_lifespan);
+  set_result += PyObject_SetAttrString(pyqos_profile, "deadline", pyqos_deadline);
+  set_result += PyObject_SetAttrString(pyqos_liveliness, "liveliness", pyqos_liveliness);
```
Comment (nburek) [inline rclpy/src/rclpy_common/src/common.c]: Should the first parameter to this function call and the one on line 109 below be `pyqos_profile`? I think that is the python object you want to set the attribute on and matches the pattern of everything above.

**Review by nburek — CHANGES_REQUESTED** [review]

File: rclpy/src/rclpy/_rclpy.c (hunk 1 above)
Comment (emersonknapp) [reply rclpy/src/rclpy/_rclpy.c]: Thanks - this is why we want to force each other to write tests for everything

**Comment by emersonknapp** [comment]
closing to open https://github.com/ros2/rclpy/pull/312 against ros2 fork

**Pull request closed without merging**